sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,458 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_RUNTIME_EPMP_H_
+#define OPENTITAN_SW_DEVICE_LIB_RUNTIME_EPMP_H_
+
+#include <stdint.h>
+
+/**
+ * RISC-V Enhanced Physical Memory Protection (EPMP).
+ *
+ * Specifications:
+ *  - PMP Enhancements for memory access and execution prevention on Machine
+ *    mode
+ *    https://docs.google.com/document/d/1Mh_aiHYxemL0umN3GTTw8vsbmzHZ_nxZXgjgOUzbvc8
+ *  - RISC-V Privileged Specfication (v. 20190608)
+ *    https://github.com/riscv/riscv-isa-manual/releases/download/Ratified-IMFDQC-and-Priv-v1.11/riscv-privileged-20190608.pdf
+ *  - Ibex PMP
+ *    https://ibex-core.readthedocs.io/en/latest/03_reference/pmp.html
+ *
+ * Assumptions (should be initialized in assembly but can be verified using
+ * `epmp_check`):
+ *   * Machine Mode Whitelist Policy is enabled (mseccfg.MMWP = 1)
+ *   * Machine Mode Lockdown is disabled (mseccfg.MML = 0)
+ */
+
+/**
+ * EPMP constants.
+ *
+ * TODO: there will probably be more of these.
+ */
+typedef enum epmp_constants {
+  kEpmpGranularity = 0,
+  kEpmpNumRegions = 16,
+} epmp_constants_t;
+
+/**
+ * Toggle for Rule Locking Bypass (mseccfg.RLB).
+ */
+typedef enum epmp_rlb_toggle {
+  kEpmpRlbToggleDisabled = 0, /**< mseccfg.RLB = 0 */
+  kEpmpRlbToggleEnabled,      /**< mseccfg.RLB = 1 */
+} epmp_rlb_toggle_t;
+
+/**
+ * EPMP state.
+ *
+ * A copy of a possible EPMP control register state.
+ */
+typedef struct epmp_state {
+  uint8_t pmpcfg[kEpmpNumRegions];
+  uintptr_t pmpaddr[kEpmpNumRegions];
+} epmp_state_t;
+
+/**
+ * EPMP entry permissions.
+ *
+ * Unlocked permissions can generally be modified when in M-mode. Locked
+ * permissions can only be modified in M-mode if Rule Locking Bypass
+ * (mseccfg.RLB) is set.
+ *
+ * When Machine Mode Lockdown is disabled (mseccfg.MML is unset) the
+ * combination R=0 W=1 is reserved. Note that this is the assumed state
+ * and so it is not possible to set these values.
+ *
+ * Note: these permissions have different meanings when Machine Mode
+ * Lockdown (mseccfg.MML) is set.
+ *
+ * TODO: MML mode could be supported by adding additional enums. That
+ * is why M-mode and S/U-mode permissions have been made explicit in
+ * enum names.
+ */
+typedef enum epmp_permissions {
+  // Unlocked permissions (can always be modified in machine mode).
+  // TODO: do we actually want to have unlocked rules? Could maybe remove
+  // these enums since MML will probably be enabled when in S/U-mode anyway.
```

Comment:
I think if this is purely for benefit of `MML`, the we can drop it. I think we should only focus on machine mode - if later we decide that we need `MML` then the library will have to be revised. This is unlikely, so I think it makes sense to leave it for now.